cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataRefreshService.java

```diff
@@ -0,0 +1,173 @@
+   */
+  @Inject
+  public SourceControlMetadataRefreshService(TransactionRunner transactionRunner,
+      NamespaceAdmin namespaceAdmin, CConfiguration cConf,
+      SourceControlOperationRunner sourceControlOperationRunner,
+      Store store) {
+    this.transactionRunner = transactionRunner;
+    this.namespaceAdmin = namespaceAdmin;
+    this.cConf = cConf;
+    this.sourceControlOperationRunner = sourceControlOperationRunner;
+    this.featureFlagsProvider = new DefaultFeatureFlagsProvider(cConf);
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    if (!isSourceControlMetadataPeriodicRefreshFlagEnabled()) {
+      return;
+    }
+    LOG.info("Starting SourceControlMetadataRefreshService");
+
+    namespaceAdmin.list().stream()
+        .map(NamespaceMeta::getNamespaceId)
+        .filter(namespaceId -> getRepositoryMeta(namespaceId) != null)
+        .forEach(this::addRefreshService);
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("Shutting down SourceControlMetadataRefreshService");
+
+    for (NamespaceSourceControlMetadataRefreshService service : refreshSchedulers.values()) {
+      service.stopAndWait();
+    }
+  }
+
+  private NamespaceSourceControlMetadataRefreshService createRefreshService(NamespaceId namespace) {
+    return new NamespaceSourceControlMetadataRefreshService(cConf, transactionRunner,
+        sourceControlOperationRunner, namespace, namespaceAdmin);
+  }
+
+  /**
+   * Adds {@link NamespaceSourceControlMetadataRefreshService} for the specified namespace. If not
+   * created before, it starts the service or else, triggers a forced manual refresh
+   *
+   * @param namespaceId The ID of the namespace for which to add the refresh service.
+   */
+  public void addRefreshService(NamespaceId namespaceId) {
+    refreshSchedulers.putIfAbsent(namespaceId, createRefreshService(namespaceId));
+    if (isSourceControlMetadataPeriodicRefreshFlagEnabled()) {
+      refreshSchedulers.get(namespaceId).start();
+      LOG.debug("Added Scheduled NamespaceSourceControlMetadataRefreshService for "
+          + namespaceId.getNamespace());
+    }
+  }
+
+  /**
+   * Stops and removes the refresh service associated with the specified namespace.
+   *
+   * @param namespaceId The ID of the namespace for which to remove the refresh service.
+   */
+  public void removeRefreshService(NamespaceId namespaceId) {
+    refreshSchedulers.computeIfPresent(namespaceId, (id, service) -> {
+      service.stop();
+      return null;
+    });
+  }
+
+  /**
+   * Retrieves the timestamp of the last refresh for the specified namespace.
+   *
+   * @param namespaceId The ID of the namespace for which to retrieve the last refresh time.
+   * @return The timestamp of the last refresh
+   */
+  public Long getLastRefreshTime(NamespaceId namespaceId) {
+    if (refreshSchedulers.containsKey(namespaceId)) {
+      return refreshSchedulers.get(namespaceId).getLastRefreshTime();
```

Comment:
This is not correct as `refreshSchedulers.get(namespaceId)` can be null. Rather get `refreshSchedulers.get(namespaceId)` in a temp variable and check if it is null